Valkey connector

bluayer opened the pull request:
<!--  Thanks for your contribution to LMCache!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/LMCache/LMCache/blob/dev/CONTRIBUTING.md
2. If this PR closes another issue, add 'Fixes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'Refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
This PR implements Valkey connector support for LMCache, addressing issue #1621.

Key implementations:
- **ValkeyConnectorAdapter**: Handles valkey:// URL parsing and connection management (standalone and cluster mode)
- **ValkeyConnector**: Using valkey-glide client with pipeline
  - Standalone : For better network latency, use mget and pipeline
  - Cluster : For better network latency, use mget, pipeline and hashslot
  - The above features should also be implemented in the Redis Connector in the future.

Performance test example :
(Random 10keys, cluster mode with 3 shards.)

- ValkeyClusterConnector PUT: 1.13ms | GET: 1.15ms | Total: 2.28ms
- RedisClusterConnector:  PUT: 6.97ms | GET: 2.00ms | Total: 8.96ms

**Special notes for your reviewers**:
Currently, it's only implemented for v1.

**If applicable**:

- [X] this PR contains user facing changes - docs added
- [X] this PR contains unit tests

bluayer committed:
feat: Add dependency for valkey-glide

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

bluayer committed:
feat: Add Valkey adapter

- Using same url prefix(valkey://) supporting standalone and cluster mode
- Use query param for checking mode.
- Standalone mode supports selecting database and cluster mode doesn't support it (but after Valkey 9, it can support database mode)

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

bluayer committed:
feat: Valkey Connector

- Implement valkey connector based on redis connector but using other library, Valkey-GLIDE managed by Valkey organization.
- Use mget and pipeline put at standalone and cluster mode
- Use Hash slot for reducing network latency at cluster mode.
- When tested with 10 keys, Valkey connector is 5 ~ 20% faster at standalone mode and up to 70%(avg 60%+) faster at cluster mode.

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

bluayer committed:
test: add test for valkey connector

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

bluayer committed:
feat: add specific version for supporting pipelinig

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

bluayer committed:
fix: linting

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

bluayer commented on tests/conftest.py:
```diff
@@ -281,6 +281,62 @@ def mock_redis_cluster():
+
+        # Process each command in the batch
+        for command_code, args in batch.commands:
+            if command_code == 1517:  # SET command
```
It's good point but there's no way to get command info when I try to use batch at valkey-glide lib.

bluayer commented on lmcache/v1/storage_backend/connector/valkey_adapter.py:
```diff
@@ -0,0 +1,141 @@
+                    url = host_part
+
+                parsed_clean = parse_remote_url(self.schema + url)
+                hosts_and_ports = [(parsed_clean.host, parsed_clean.port)]
+
+            return ValkeyClusterConnector(
+                hosts_and_ports=hosts_and_ports,
+                username=username,
+                password=password,
+                loop=context.loop,
+                local_cpu_backend=context.local_cpu_backend,
+            )
+        else:
+            # Standalone mode - use standalone connector
+            # Local
+            from .valkey_connector import ValkeyConnector
+
+            logger.info("Creating ValkeyConnector for standalone mode")
+
+            # Parse auth info and database for standalone
+            username, password = self._parse_auth_from_url(url)
+            database_id = self._parse_database_from_url(url)
+
+            # Remove auth info: user:pass@host:port/db -> host:port/db
+            if "@" in url:
+                _, host_part = url.split("@", 1)
+                url = host_part
+
+            # Remove database: host:port/db -> host:port
+            if "/" in url:
+                url = url.rsplit("/", 1)[0]
+
+            return ValkeyConnector(
+                url=url,
+                loop=context.loop,
+                local_cpu_backend=context.local_cpu_backend,
+                username=username,
+                password=password,
+                database_id=database_id,
+            )
```
I some agree, because the types of parameters the connector can receive are similar yet different.

bluayer committed:
fix: add checking memory equality at test

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

bluayer commented on tests/v1/test_connector.py:
```diff
@@ -238,6 +238,115 @@ def test_redis_connector(url, autorelease_v1):
+
+    # Test 5: Retrieve and verify data
+    future = asyncio.run_coroutine_threadsafe(connector.get(random_key), async_loop)
+    future.result()
```
Good point. I fixed it.

YaoJiayi commented:
@maobaolong Would you like to take a look at this PR?

bluayer commented:
Hi @maobaolong, This PR is quite large, and I know you're busy, but would you review it please?

maobaolong commented:
@YaoJiayi Thanks for invite me to review this PR, and thanks also for your efforts to pick up the valkey connector. @bluayer 

Will take a look this week. I'm sorry that this week we are busy for some urgency work.

maobaolong commented:
@bluayer First of all, may i ask whether you use valkey enable RDMA? Or just use tcp mode?

bluayer commented:
@maobaolong Thank you for your response. 

Only use tcp in this PR. Because the main client used and developed by valkey org, valkey-glide, does not yet support RDMA. (Ref : https://github.com/valkey-io/valkey-glide/issues/1963)

I plan to work on adding RDMA support after merging this connector.

maobaolong commented on requirements/common.txt:
```diff
@@ -14,6 +14,7 @@ py-cpuinfo
 pyyaml
 pyzmq >= 25.0.0
 redis
+valkey-glide>=2.0.0
```
It is not necessary to add valkey requirement into lmcache `requirement` by default.

This is what the design of pluggable remote connector framework, the main idea is that we do not need to install all the dependencies of connectors, instead, user can install the relevant requirement separately. 

If you want, maybe you can add a connector-valkey.txt within `valkey-glide>=2.0.0`.

bluayer replied:
I understand it, but why is Redis included in common.txt? If it's designed as you said, then Redis should also be deleted from common.txt.

maobaolong replied:
@bluayer Redis is used by another place other than remote connector.

sammshen replied:
@bluayer, redis was previously used as a p2p connector. I think you can include instructions on how to install it in the docs :)

maobaolong commented on lmcache/v1/storage_backend/connector/valkey_adapter.py:
```diff
@@ -0,0 +1,141 @@
+    def _parse_database_from_url(self, url: str) -> Optional[int]:
+        database_id: Optional[int] = None
+        if "/" in url:
+            url, db_part = url.split("/", 1)
```
It really complex for the parameters parse work. it is not introduced by this PR.

But you may reference other connector implementation and leverage the `extra_config`, it would be much easy to pass some extra_config belongs to `valkey`, something like `valkey_password`, `valkey_username`, `valkey_database_id`.

@sammshen What about your idea?

bluayer replied:
I referenced the RedisAdapter, and I agree that it's complex. Using `extra_config` is a good approach.

sammshen replied:
I think both work. just add some clear docs if you are going to put it into the url

maobaolong commented on lmcache/v1/storage_backend/connector/valkey_connector.py:
```diff
@@ -0,0 +1,363 @@
+                "Key exists but KV cache does not exist."
+                "Might happen when the cache is evicted by valkey."
+            )
+            return None
```
This is possible to lead memory leak, you have to add `memory_obj.ref_count_down()`

@ApostaC @YaoJiayi @sammshen This is a case what the https://github.com/LMCache/LMCache/issues/1713 hope to defeat.

sammshen replied:
I think the ref count down is done in storage manager but I agree this is very confusing

maobaolong commented on lmcache/v1/storage_backend/connector/valkey_connector.py:
```diff
@@ -0,0 +1,363 @@
+            converted = bytes(kv_bytes)
+            view[: metadata.length] = converted
+
+        return memory_obj
```
You have to add your main code of this method surrounded by `try except` code block, so that whatever exception occurred, the allocated memory object can ref_count_down.

maobaolong reviewed (changes requested):
@bluayer Left some comments, about the `turn url parse to extra_config` suggestion, let's wait the feedback from @sammshen .

sammshen commented:
@bluayer, do you have any benchmarks on the read and write throughput (in terms of GB / s) of Valkey? (any benchmarking utility is fine)? :)

Also, would you mind updating the docs? https://docs.lmcache.ai/kv_cache/storage_backends/valkey.html

sammshen reviewed (approved):
LGTM on first pass! Please add documentation

bluayer committed:
feat: Use extra config for simple paramters

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

bluayer committed:
fix: remove valkey-glide depedency from common.txt

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

bluayer committed:
feat: Add try-except and prevent memory leak

- Add try-except under assert statement for checking error
- Add memory_obj.ref_count_down() for preventing memory leak

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

bluayer committed:
feat: double check for config and extra_config

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

bluayer committed:
fix: lint

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

bluayer committed:
docs: Add docs for valkey connector

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

bluayer commented:
Hi @maobaolong, I've applied all of your feedback. Please check it out :) Also I added the document for valkey connector.
And due to dependencies on Valkey-GLIDE, the unit tests are failing. How can I resolve this? Please share your ideas.

And @sammshen, I don't have my own benchmark results but you can estimate throughput by this official blog post(https://valkey.io/blog/unlock-one-million-rps/). Valkey 7.2 is completely identical to Redis 7.2, so this high performance has been supported since Valkey 8 released last year.

sammshen commented:
you can mock the behavior in the unit tests

bluayer committed:
fix: Move import statement to the head of file and exception handling

Signed-off-by: Jungwoo Song <bluayer@gmail.com>

maobaolong commented on lmcache/v1/storage_backend/connector/valkey_connector.py:
```diff
@@ -0,0 +1,382 @@
+                converted = bytes(kv_bytes)
+                view[: metadata.length] = converted
+        except Exception as exc:
+            logger.error(f"Fail to converting : {exc}")
```
@bluayer The way you swallow the exception, it mean that the caller will not aware this exception occurred, suggest to return None.

bluayer replied:
I added `return None` at recent commit.

maobaolong reviewed (commented):
@bluayer Left a comment within get method.

maobaolong commented on lmcache/v1/storage_backend/connector/valkey_connector.py:
```diff
@@ -0,0 +1,386 @@
+        async def create_connection():
+            try:
+                # Third Party
+                from glide import GlideClientConfiguration, ServerCredentials
```
Why we import these here, rather than import in the head of this file?

bluayer replied:
Initially, I placed import statements there to handle `import exceptions`, but as you suggested in your feedback, now moving it to the file's head seems better. I've made the change.

maobaolong commented on lmcache/v1/storage_backend/connector/valkey_connector.py:
```diff
@@ -0,0 +1,386 @@
+        async def create_connection():
+            try:
+                # Third Party
+                from glide import GlideClusterClientConfiguration, ServerCredentials
```
Maybe we can move this line into the head of this file, wdyt?

maobaolong reviewed (approved):
@bluayer Left a minor comment inline. Otherwise, LGMT.

sammshen commented:
please merge from https://github.com/LMCache/LMCache/pull/1833#issuecomment-3424616997 to fix integration tests

bluayer commented:
@sammshen I pull dev branch and rebase this branch. Thank you.

sammshen commented:
thanks @bluayer for the PR and @maobaolong for the great feedback!

meet-bhagdev commented:
this is awesome! thank you @bluayer
